docs/_docs/05_rfc/variable_output_modification.md

```diff
@@ -0,0 +1,152 @@
+
+# NOTE: this must be marked as sensitive since it is derived from a sensitive variable
+output "my_password_hashed" {
+  value = base64sha256(var.my_password)
+}
+
+output "length_my_list" {
+  value = length(var.my_list)
+}
+```
+
+Using Terragrunt with this module will run into the following issue:
+
+- Because the output `my_password_hashed` is not marked as sensitive, terraform will error out.
+- `my_list` is missing the type definition, so the input from `terragrunt` will be interpretted as a string. This means
+  that the output `length_my_list` will be the string length, and not the list length.
+
+We will need to transform these variables and outputs. We will introduce the `transform` block to handle this. The
+following `terragrunt.hcl` configuration indicates the necessary transformations to `variable` and `output` to support
+deployment:
+
+```hcl
+transform {
+  # Sub blocks are 'variable' or 'output', to indicate what terraform block needs to be transformed. Next, the label
+  # should match with the corresponding variable label defined in the terraform module.
+  # Each subblock matches the underlying terraform block, and under the hood, the attributes and blocks are merged into
+  # the terraform module.
+  # For example, in this first variable subblock, the 'type' attribute is merged into the upstream terraform module
+  # before terragrunt invokes terraform.
+  variable "my_list" {
+    type = list(string)
+  }
+  output "my_password_hashed" {
+    sensitive = true
+  }
+}
+```
+
+As indicated in the comment, Terragrunt will merge these `variable` and `output` configurations into the module code
+prior to invoking Terraform. That is, the following will happen when `terragrunt apply` is invoked:
+
+- Terragrunt parses the configurations. The transform operations are interpretted and recorded internally at this point.
+- Terragrunt clones the module source into the working directory (`.terragrunt-cache`).
+- `generate` blocks are processed and copied into the working directory.
+- `transform` blocks are processed. In this stage, Terragrunt will scan the `variable` and `output` blocks in the
+  underlying module cloned in the working directory, and **directly modify the local source** with the updates. In this
+  case, the `my_list` variable block definition will have the `type = list(string)` attribute set, and the
+  `my_password_hashed` output will have `sensitive = true` attribute set.
+- Terragrunt invokes `terraform apply` on the modified module.
+
+In this way, Terragrunt can convert the underlying shared module into a root module that can be deployed directly.
+Under the hood, this transformation is implemented in the same way that the `aws-provider-patch` command works. This
+should be feasible by doing the following:
+
+- Scan all `.tf` files in the directory.
+- For each file found, parse using the `hclwrite` parser.
+- Walk the AST, looking for `variable` or `output` blocks that match the `transform` sub blocks.
```

Comment:
Theoretically, this could also rewrite resources? 🤯

Reply:
Yes in theory, although I think that would be fairly complex. I think at that point, forking the underlying module is a better approach.